proposals/pattern-const-expressions.md

```diff
@@ -0,0 +1,145 @@
+[semantics]: #semantics
+A pattern expression only consisting of constant subpatterns with a constant LHS may be evaluated at compile time currently, but it is not considered as constant. With this change, the following adjustments enable constant pattern expressions to be used in constant contexts, like assigning a constant field or local, a default parameter, the condition in a constant conditional expression, etc.
+
+### Examples
+[examples]: #examples
+The below example shows pattern expressions being assigned to constant fields:
+```csharp
+public const int A = 4;
+public const bool B = A is 4;
+public const bool C = A is not 3 and <= 4 or 6;
+
+// DeploymentEnvironment is an enum type
+public const DeploymentEnvironment Environment = DeploymentEnvironment.Production;
+public const bool D = Environment
+    is DeploymentEnvironment.Production
+    or DeploymentEnvironment.Test;
+```
+
+In the expression assigned to `C`, since `A` is constant, and all the subpatterns on the pattern are constant, the entire expression only consists of constants. This result will be evaluated during compile-time and then be assigned to the constant field.
+
+Likewise, `Environment` is constant as an enum value, and so are the other subpatterns of the expression assigned to `D`.
+
+Another example, using `null` and default value checks:
+```csharp
+const int a = 4;
+const bool b = false;
+const long c = 4;
+const string d = "hello";
+
+const bool x = a is default(int); // always false
+const bool y = b is default(bool); // always true
+const bool z = d is not null; // always true
+const bool p = b is null; // always false
+```
+
+All the above are currently valid pattern matching expressions, that also emit warnings about their constant evaluation results, about them being always true or false.
+
+When assigning those expressions to a constant symbol, these warnings about the constant result of the expression will **not** be reported, as the user intends to capture the constant value of the expression.
+
+Pattern expressions containing non-constant subpatterns, like accessing properties, list patterns and var patterns, are **not** constant. In the below examples, all expressions will report compiler errors:
+
+```csharp
+const bool q = d is string { Length: 5 }; // Error: not a constant expression
+const bool r = d is [.. var prefix, 'l', 'o']; // Error: not a constant expression
+const bool s = d is var someString; // Error: not a constant expression
+```
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+None.
+
+## Alternatives
+[alternatives]: #alternatives
+
+Currently, equality and comparison operators can be used to compare against other literals, including nullability of the objects (e.g. `x != null`, or `y == 4 || y < 3`).
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- [ ] Requires LDM review
+- [ ] Should this be extended to `switch` expressions too? (Discussion [#7489](https://github.com/dotnet/csharplang/discussions/7489))
```

Comment:
LDM did seem in favor of this, but I would hold off on fully specifying until we work through the subsumption warnings and errors issues.